Makefile

```diff
@@ -0,0 +1,52 @@
+MAKEFLAGS += --no-builtin-rules
+MAKEFLAGS += --no-builtin-variables
+
+# make tasks runn all commands in a single shell
+.ONESHELL:
+
+# these targets don't produce files:
+.PHONY: clean filecheck pytest tests rerun-notebooks precommit-install precommit black pyright
+
+# remove all caches and the venv
+clean:
+	rm -rf venv .pytest_cache *.egg-info .coverage.*
+
+# run filecheck tests
+filecheck:
+	lit -vv tests/filecheck
+	lit -vv docs/Toy/examples
+
+# run pytest tests
+pytest:
+	pytest tests -W error -vv
+
+# run all tests
+tests: pytest filecheck pyright
+	echo test
+
+# re-generate the output from all jupyter notebooks in the docs directory
+rerun-notebooks:
+	jupyter nbconvert --ClearMetadataPreprocessor.enabled=True --inplace --to notebook --execute docs/*.ipynb
+
+# set up all precommit hooks
+precommit-install:
+	pre-commit install
+
+# run all precommit hooks and apply them
+precommit:
+	pre-commit run --all
+
+# run pyright on all files in the current git commit
+pyright:
+	pyright $(shell git diff --staged --name-only)
+
+# run black on all files currently staged
+black:
+	black $(shell git diff --staged --name-only)
```

Comment:
I think this fails because you have nothing staged, and this target only runs black on things that are staged.

We can expand this to run on the whole thing is nothing is staged

Reply:
would be good if nothing failed tbh

Reply:
Ah yeah, that makes sense. Is there a (reasonable) way to make this target work without anything staged?

Reply:
it's arbitrary bash, so probably yes^^

Reply:
NVM, let's forget about it. No need to overengineer somethign for such a minor issue.

Reply:
Done!

Reply:
Nice, works 👍